airflow/ti_deps/deps/branch_dep.py

```diff
@@ -0,0 +1,63 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from airflow.ti_deps.deps.base_ti_dep import BaseTIDep
+
+
+class BranchDep(BaseTIDep):
+    """
+    Determines if any of the task's direct upstream relatives have decided this task should
+    be skipped.
+    """
+    NAME = "Branch Rule"
+    IGNORABLE = True
+    IS_TASK_DEP = True
+
+    def _get_dep_statuses(self, ti, session, dep_context=None):
+        from airflow.models.skipmixin import SkipMixin
+        from airflow.utils.state import State
+
+        upstream = ti.task.get_direct_relatives(upstream=True)
```

Comment:
In your example, if I clear false_2 it would re run -- the skip task was not a direct relative. That makes this behaviour even more confusing as sometimes it would run and sometimes it wouldn't depending on which task was cleared.

Reply:
I did have the same concern as you. So I did some experiment. If someone clears `false_2`, Airflow actually does the right thing. It becomes skipped. That's already happening in Airflow, not because of my PR.
The doc describes this behaviour as "skipped states propagates". It's actually `trigger_rule_deps.py` that does this "magic". Somewhere here the state is set to SKIPPED if any of its upstream is skipped and the task `trigger_rule` is `all_success`. This is exactly what we hope for.

```LANG=python
            if tr == TR.ALL_SUCCESS:
                if upstream_failed or failed:
                    ti.set_state(State.UPSTREAM_FAILED, session)
                elif skipped:
                    ti.set_state(State.SKIPPED, session)
```

Reply:
Ah great! I wonder if that behaviour is tested already or not

Reply:
No I can't find a test for this useful behaviour. So I'm adding one.